cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/executor/RemoteDatasetOpExecutor.java

```diff
@@ -64,6 +76,34 @@ public EndpointStrategy get() {
   }
 
   @Override
+  protected void startUp() throws Exception {
+    // wait for dataset executor to be discoverable
+    LOG.info("Starting DatasetOpExecutor.");
+    int timeout = cConf.getInt(Constants.Startup.STARTUP_SERVICE_TIMEOUT);
+    try {
+      Tasks.waitFor(true, new Callable<Boolean>() {
+        @Override
+        public Boolean call() throws Exception {
+          return endpointStrategySupplier.get().pick() != null;
+        }
+      }, timeout, TimeUnit.SECONDS, Math.min(timeout, Math.max(10, timeout / 10)), TimeUnit.SECONDS);
+      LOG.info("DatasetOpExecutor started.");
+    } catch (TimeoutException e) {
+      // its not a nice message... throw one with a better message
+      throw new RuntimeException(String.format("Timed out waiting to discover the %s service. " +
+                                                 "Check the container logs then try restarting the service.",
+                                               Constants.Service.DATASET_EXECUTOR));
+    } catch (InterruptedException e) {
+      throw new RuntimeException(String.format("Interrupted while waiting to discover the %s service.",
+                                               Constants.Service.DATASET_EXECUTOR));
+    } catch (ExecutionException e) {
+      throw new RuntimeException(String.format("Error while waiting to discover the %s service. Root cause: %s",
```

Comment:
In this case, we ran into an unexpected exception, right? I think the complete stack trace would be useful, in addition to the root cause.